### ISSUE TITLE
Zuul removes content-length for all clients, it is not a hop-by-hop header

### ISSUE THREAD

**Issue by jadetang** [spring-cloud/spring-cloud-netflix#3008]
In the method `isIncludedHeader`  of `ProxyRequestHelper`, the code will not pass some headers, any reason to do this?

`

	public boolean isIncludedHeader(String headerName) {
		String name = headerName.toLowerCase();
		RequestContext ctx = RequestContext.getCurrentContext();
		if (ctx.containsKey(IGNORED_HEADERS)) {
			Object object = ctx.get(IGNORED_HEADERS);
			if (object instanceof Collection && ((Collection<?>) object).contains(name)) {
				return false;
			}
		}
		switch (name) {   //these header will not be passed to the downstream server.
		case "host":
		case "connection":
		case "content-length":
		case "content-encoding":
		case "server":
		case "transfer-encoding":
		case "x-application-context":
			return false;
		default:
			return true;
		}
	}
`

**Comment by spencergibb** [comment]
Proxies are supposed to remove some headers. https://tools.ietf.org/html/draft-ietf-httpbis-p1-messaging-14#section-7.1.3

**Comment by jadetang** [comment]
@spencergibb correct me if I am wrong. It only says should remove `connection` header

**Comment by spencergibb** [comment]
The following HTTP/1.1 header fields are hop-by-hop header fields:

   o  Connection

   o  Keep-Alive

   o  Proxy-Authenticate

   o  Proxy-Authorization

   o  TE

   o  Trailer

   o  Transfer-Encoding

   o  Upgrade

**Comment by jadetang** [comment]
but according to the code, it remove  `content-length` as well.

**Comment by spencergibb** [comment]
I think that is a relic of apache httpclient.

**Comment by jadetang** [comment]
no, this class if from spring-zuul, maybe I should update the version? It is `Brixton.SR7`

**Comment by ryanjbaxter** [comment]
@jadetang I am slightly confused now.  At first, this issue was complaining that `content-length` was removed but now you are talking about `content-encoding` (as well as in #3059).  Are you convinced that is we dont remove `content-encoding` that is will fix your issue?  (#3062 will only keep `content-encoding`, `content-length` will still be removed)

**Comment by jadetang** [comment]
@ryanjbaxter 
Sorry for confusing you. 
First of all, both the `content-encoding` and the `content-length` should not be removed since they are not hopper-by-hopper headers. 
Second, the `ProxyRequestHelper` tries to remove the `content-length`, but in the `SimpleHostRoutingFilter`, the `content-length` is set by this line later
`		InputStreamEntity entity = new InputStreamEntity(requestEntity, contentLength, contentType);
`
The #3062 works for me because of that. But in general, I think we need look to deeper for this, why remove some headers in the first place.

**Comment by jadetang** [comment]
@ryanjbaxter yes, I think so.

**Comment by ryanjbaxter** [comment]
I would rather just take care of both in on PR.

**Comment by mdiniz** [comment]
@ryanjbaxter I gave it a thought and I think no change related to `Content-Length` is needed, the confusion lies on responsibility of the method `ProxyRequestHelper.isIncludedHeader` called by `ProxyRequestHelper.buildZuulRequestHeaders`.
Both existing routers `SimpleHostRoutingFilter` and `RibbonRoutingFilter` call this method. That means they get a Map without `Transfer-Encoding` (which we all agree is correct) and without `Content-Length` (which looks incorrect). Then each of these filters is responsible for getting the content length from the incoming request and setting it on the outgoing request (or they could set `Transfer-Encoding`, but they don't). So the outgoing requests have the correct headers.
If we don't strip `Content-Length` on `ProxyRequestHelper.isIncludedHeader`, then these filters (and any other out there) will add a second `Content-Length` header on the outgoing request. I tested not stripping the header and running the test suite, resulting in Failures: 36, Errors: 13. The 13 errors on `SimpleHostRoutingFilterTests` are about a duplicated `Content-Length` header.
The confusion is that at first sight, when `ProxyRequestHelper.isIncludedHeader` removes a header, it won't/shouldn't be sent. However, the method merely prepares the landscape so that the router filters can set extra headers as needed.
